Add details of StatementCouldNotBeMapped in errors.md (#185)
cipherstash/proxy: docs/errors.md

<!-- ---------------------------------------------------------------------------------------------------- -->

## Statement could not be type checked <a id='mapping-statement-could-not-be-type-checked'></a>

An error occurred when attempting to type check the SQL statement.

### Error message

```
Statement could not be type checked: '{type-check-erro-message}'
```

### Notes

CipherStash Proxy checks SQL statements against the database schema to transparently encrypt and decrypt data.

The behaviour of Proxy depends on the `mapping_errors_enabled` configuration.

> When `mapping_errors_enabled == false` (the default) type check errors are logged, and the statement is passed through to the database.

Reviewer: ```suggestion
When `mapping_errors_enabled` is `false` (the default), then type check errors are logged, and the statement is passed through to the database.
```

> When `mapping_errors_enabled == true` type check errors are raised and statement execution halts.

Reviewer: ```suggestion
When `mapping_errors_enabled` is `true`, then type check errors are raised, and statement execution halts.
```

In our experience, most production systems have a relatively small number of columns that require protection.

> As SQL is large and complex, rather than block statements with false negative type check errors, the default behaviour is to allow the statement.

Reviewer: ```suggestion
As SQL is large and complex, instead of blocking statements with type check errors that are false negatives, the default behaviour of Proxy is to allow the statement.
```

> However, this does mean it is possible that a statement referencing encrypted columns cannot be type-checked and is passed through to the database.

Reviewer: ```suggestion
However, this does mean it is possible that a statement that references encrypted columns cannot be type-checked, and it will be passed through to the database.
```

> In this case, database column constraints (included in EQL) will catch the statement and return a PostgreSQL error.

Reviewer: ```suggestion
When a statement is passed through to the database, the database's column constraints (provided by EQL) will catch the statement, and return a PostgreSQL error.
```

Example constraint error:
```sql
ERROR:  Encrypted column missing version (v) field: 34234
CONTEXT:  PL/pgSQL function _cs_encrypted_check_v(jsonb) line 6 at RAISE
SQL function "cs_check_encrypted_v1" statement 1
```

### How to Fix

In most cases, this error will occur if the statement contains invalid or unsupported syntax.

> Check that the installed version of CipherStash Proxy, and update to the latest version if possible.

Reviewer: ```suggestion
Check if you are running the latest version of CipherStash Proxy, and update to the latest version if not.
```

If the error persists, please contact CipherStash [support](https://cipherstash.com/support).

<!-- ---------------------------------------------------------------------------------------------------- -->

## Internal Mapper error <a id='mapping-internal-error'></a>

An internal error occurred when attempting to type check or transform a SQL statement.

> This could be due to an internal invariant failure or because of a specific fragment of unsupported SQL syntax.

Reviewer: ```suggestion
This could be due to an internal invariant failure, or because of a specific fragment of unsupported SQL syntax.
```

### How to Fix

> Check that the installed version of CipherStash Proxy, and update to the latest version if possible.

Reviewer: ```suggestion
Check if you are running the latest version of CipherStash Proxy, and update to the latest version if not.
```
